Adding Layer Ablation

vivekmig opened the pull request:
This PR adds layer ablation, which computes the attribution of the input / output of a layer, by setting each neuron or group of neurons to a given baseline and measuring the difference in output value. This is analogous to feature ablation for layer attribution, with baselines and feature masks given with respect to the layer input / output size.

Differential Revision: D18670637

NarineK commented:
Let's add a brief description to this one and Neuron Ablation as well.

NarineK commented on captum/attr/_core/layer/layer_ablation.py:
```diff
@@ -0,0 +1,273 @@
+            >>>                          layer_mask=layer_mask)
+        """
+
+        def layer_forward_func(*args):
```
nit: could this be passed in 2 arguments, like ?
```
layer_input =inputs
original_inputs = additional_forward_args
```

vivekmig replied:
This would work if layer_input is guaranteed to be 1 tensor / element as it is now, but once that becomes variable, I'm not sure how to cleanly separate it, since _run_forward in feature ablation will concatenate the inputs and additional_args tuples?

vivekmig commented on captum/attr/_core/layer/layer_ablation.py:
```diff
@@ -0,0 +1,273 @@
+
+            all_layer_inputs = {}
+            if device_ids is not None:
+                scattered_layer_input, _ = scatter_kwargs(
```
Good point, will update!

vivekmig commented on captum/attr/_core/layer/layer_ablation.py:
```diff
@@ -0,0 +1,273 @@
+                and isinstance(self.forward_func, torch.nn.DataParallel)
+                and self.forward_func.device_ids is not None
+            ):
+                device_ids = self.forward_func.device_ids
```
Good point, the getattr syntax is shorter! I think it might be good to still leave the DataParallel instance check though, it's possible a model might have that attribute and not be DataParallel / not need to be scattered?

vivekmig commented on captum/attr/_core/layer/layer_ablation.py:
```diff
@@ -0,0 +1,273 @@
+
+        return ablator.attribute(
+            layer_eval,
+            baselines=baselines,
```
Makes sense, yeah, it's a different here from baselines in other algorithms. I think since we want to evaluate essentially "turning off" each layer input / output, it's not clear what the layer values would be for a given input baseline. Maybe we can document the difference between this baseline and others more clearly? Or rename the argument to something like layer_baseline instead?

vivekmig commented:
Thanks for the review! Addressed most of the comments.
1. I've renamed to LayerFeatureAblation and also refer to ablating groups of layer input / output elements. I think the confusion in naming is based on defining each neuron as a unique set of weights as opposed to each scalar in the layer as an input / output "neuron", so hopefully this is clearer!

2. Makes sense, yes, this isn't completely consistent with baselines for Layer IG and Conductance, but I do think it's preferable to keep it this way here, since we want to essentially "turn off" each layer input / output neuron, and that's more difficult with an input baseline. We could rename the baselines argument to layer_baselines to make the distinction clearer, what do you think?

NarineK commented on captum/attr/_core/layer/layer_feature_ablation.py:
```diff
@@ -0,0 +1,270 @@
+                all_layer_inputs[layer_input.device] = layer_input
+
+            def forward_hook(module, inp, out=None):
+                if inp[0].device not in all_layer_inputs:
```
I moved `_extract_device` to common - we could use it here too ? https://github.com/pytorch/captum/pull/222/files#diff-3ff60ae8a182cf817df47a82964d8730R438

vivekmig replied:
Sounds good, I will wait for that PR to land and then update

NarineK commented on captum/attr/_core/layer/layer_feature_ablation.py:
```diff
@@ -0,0 +1,270 @@
+            if device_ids is None and isinstance(
+                self.forward_func, torch.nn.DataParallel
+            ):
+                device_ids = getattr(self.forward_func, "device_ids", None)
```
I'm not able to comment in our previous conversation thread about the `device_ids ` - I'll comment here:
I think currently we have some places where we check with `torch.nn.DataParallel` and some cases where we don't (e.g in `gradient.py`).
If we don't explicitly check for `torch.nn.DataParallel` then it is more generic but the user might not always want it as you mentioned. 
But from other point of view they can always set `self.forward_func.device_ids=None` if they don't want the computations to be performed in a distributed manner ?

vivekmig replied:
Makes sense, sounds good, it would be good to choose one approach and stay consistent. Should we not check for torch.nn.DataParallel everywhere when accessing device_ids and update documentation to reflect this?

NarineK replied:
Yeah, probably that would be good - in case we want to also support ModelParallel, DistributedDataParallel in the future and custom forward functions.

vivekmig replied:
Sounds good
